### DISCUSSION TITLE
Use network namespaces for each GTP net-device instead of nr-bind / libdevbnd.so

### DISCUSSION THREAD

**Discussion by laf0rge** [aligungr/UERANSIM#728] [Ideas]
The approach taken by nr-bind / libdevbnd with a LD_PRELOAD wrapper is a 1990s-style approach of trying to address the problem of making sure one application uses one specific GTP net-device.  It has many problems, as the IP stack and the routing table is not isolated from the rest of the system.  Also, LD_PRELOAD will not work for statically linked binaries like commonly found in the rust and go language world.

An IMHO much more superior mode would be to create one separte network namespace for each GTP tunnel.  That network namespace is completely isolated from the host/root networking.  It has its own IP adresses, routing table, etc.  You can then start arbitrary programs inside that network namespace, and they will behave truly like the application would run on a computer whose only network interface is a 5G cellular modem.

This approach has been implemented (for example) in [osmo-uecups](https://projects.osmocom.org/projects/osmo-uecups/wiki).

**Comment by aligungr** [comment]
Yes, this is the correct approach for UE network connection simulation.

For anyone interested in this topic, pull requests are welcome :)
